### PR TITLE
Support for 3DS Friends

### PR DESCRIPTION
**Pull request by hauntii**
This PR adds support for most methods of Friends for the 3DS.

Methods:

- [X] UpdateProfile
- [X] UpdateMii
- [ ] UpdateMiiList
- [ ] UpdatePlayedGames
- [X] UpdatePreference
- [X] GetFriendMii
- [ ] GetFriendMiiList
- [ ] IsActiveGame
- [X] GetPrincipalIDByLocalFriendCode (Stubbed)
- [ ] GetFriendRelationships
- [X] AddFriendByPrincipalID
- [X] RemoveFriendByLocalFriendCode (Stubbed)
- [X] RemoveFriendByPrincipalID
- [X] GetAllFriends
- [ ] UpdateBlacklist
- [X] SyncFriend
- [X] UpdatePresence
- [X] UpdateFavoriteGameKey
- [X] UpdateComment
- [ ] UpdatePicture
- [X] GetFriendPresence
- [X] GetFriendComment
- [ ] GetFriendPicture
- [X] GetFriendPersistentInfo
- [ ] SendInvitation

All methods added have been extensively tested. Local and Internet friends work correctly.

The global for `Presence` has been renamed to `PresenceV2`, and a Presence global has been added in its place for 3DS presence. 

The README has been updated to reflect the current status of the server as well.

Notifications that do not have an associated file have been placed in a `notifications.go` file in their respective platform folders.

Here are some demo videos:

<details>
  <summary>Demo Videos</summary>

Internet

https://user-images.githubusercontent.com/57009359/221081053-0e3168a6-a7a7-4dec-a181-74433a042c3a.mp4

Local

https://user-images.githubusercontent.com/57009359/221081063-528aa3ab-777f-4870-9799-ad17bc4a2a3c.mp4

Presence

https://user-images.githubusercontent.com/57009359/221081077-38057155-4cd2-4831-8bd9-a816dfb99b5e.mp4

</details>

### CODE REVIEW

File: assign_nex_protocols.go [hunk 1]
```diff
@@ -36,9 +36,17 @@ func assignNEXProtocols() {
 	// Friends (3DS) protocol handles
 	friends3DSServer.UpdateProfile(friends_3ds.UpdateProfile)
 	friends3DSServer.UpdateMii(friends_3ds.UpdateMii)
-	friends3DSServer.UpdatePreference(friends_3ds.UpdatePreference)
+	friends3DSServer.UpdatePreference(friends_3ds.UpdatePreferences)
```
Comment (jonbarrow) [inline assign_nex_protocols.go]: `UpdatePreference` is the real method name. Why was this renamed to `UpdatePreferences`? Function names should mirror the official method names

File: 3ds/get_pid_by_lfc.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/PretendoNetwork/friends-secure/globals"
+	nex "github.com/PretendoNetwork/nex-go"
+	nexproto "github.com/PretendoNetwork/nex-protocols-go"
+)
+
+func GetPrincipalIDByLocalFriendCode(err error, client *nex.Client, callID uint32, lfc uint64, lfcList []uint64) {
+	// Drop this request entirely, waiting for gRPC to retrieve PID from account server
+
+	rmcResponse := nex.NewRMCResponse(nexproto.Friends3DSProtocolID, callID)
+	rmcResponse.SetSuccess(nexproto.Friends3DSMethodGetPrincipalIDByLocalFriendCode, nil)
```
Comment (jonbarrow) [inline 3ds/get_pid_by_lfc.go]: This should probably be an RMC error using the `Unimplemented` error code, unless the success is required for something else?

File: 3ds/notifications.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package friends_3ds
+
+import (
+	database_3ds "github.com/PretendoNetwork/friends-secure/database/3ds"
+	"github.com/PretendoNetwork/friends-secure/globals"
+	nex "github.com/PretendoNetwork/nex-go"
+	nexproto "github.com/PretendoNetwork/nex-protocols-go"
+)
+
+// Notifications that are used in other files with no main file
```
Comment (jonbarrow) [inline 3ds/notifications.go]: Maybe _all_ notifications should be placed in their own files? For example putting `SendUserWentOfflineNotificationsGlobally` into `send_user_went_offline_notification.go` as simple `SendUserWentOffline`. Or at least a `notifications.go` file for each console? That feels cleaner than having some notification functions in a globals file and some in other files, and Go convention is to keep one file per function anyway?

File: 3ds/remove_friend_by_lfc.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package friends_3ds
+
+import (
+	"github.com/PretendoNetwork/friends-secure/globals"
+	nex "github.com/PretendoNetwork/nex-go"
+	nexproto "github.com/PretendoNetwork/nex-protocols-go"
+)
+
+func RemoveFriendByLocalFriendCode(err error, client *nex.Client, callID uint32, friendLFC uint64) {
+	// Drop this request entirely, waiting for gRPC to retrieve PID from account server
```
Comment (jonbarrow) [inline 3ds/remove_friend_by_lfc.go]: This should probably be an RMC error using the `Unimplemented` error code, unless the success is required for something else?

File: 3ds/update_preferences.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/PretendoNetwork/friends-secure/globals"
+	nex "github.com/PretendoNetwork/nex-go"
+	nexproto "github.com/PretendoNetwork/nex-protocols-go"
+)
+
+func UpdatePreferences(err error, client *nex.Client, callID uint32, showOnline bool, showCurrentGame bool, showPlayedGame bool) {
+	if !showCurrentGame {
+		emptyPresence := nexproto.NewNintendoPresence()
+		emptyPresence.GameKey = nexproto.NewGameKey()
+		emptyPresence.ChangedFlags = 4294967295 // FF FF FF FF, All flags
```
Comment (jonbarrow) [inline 3ds/update_preferences.go]: Large numbers like this should probbaly be represented as hex for easier ready. `0xFFFFFFFF` instead of 4294967295

File: 3ds/update_presence.go [hunk 6]
```diff
@@ -1,13 +1,24 @@
...
-func UpdatePresence(err error, client *nex.Client, callID uint32, presence *nexproto.NintendoPresence, unknown bool) {
-	// TODO: Do something with this
+func UpdatePresence(err error, client *nex.Client, callID uint32, presence *nexproto.NintendoPresence, showGame bool) {
+	currentPresence := presence
+
+	// Send an entirely empty status, with every flag set to update
+	if !showGame {
+		currentPresence = nexproto.NewNintendoPresence()
+		currentPresence.GameKey = nexproto.NewGameKey()
+		currentPresence.ChangedFlags = 4294967295 // FF FF FF FF, All flags
```
Comment (jonbarrow) [inline 3ds/update_presence.go]: Large numbers like this should probbaly be represented as hex for easier ready. `0xFFFFFFFF` instead of 4294967295

File: 3ds/update_presence.go [hunk 7]
```diff
@@ -1,13 +1,24 @@
...
 	nex "github.com/PretendoNetwork/nex-go"
 	nexproto "github.com/PretendoNetwork/nex-protocols-go"
 )
 
-func UpdatePresence(err error, client *nex.Client, callID uint32, presence *nexproto.NintendoPresence, unknown bool) {
-	// TODO: Do something with this
+func UpdatePresence(err error, client *nex.Client, callID uint32, presence *nexproto.NintendoPresence, showGame bool) {
+	currentPresence := presence
+
+	// Send an entirely empty status, with every flag set to update
```
Comment (jonbarrow) [inline 3ds/update_presence.go:13]: Do we know what the other flags are?

File: nex_server.go [hunk 8]
```diff
@@ -30,27 +33,28 @@ func startNEXServer() {
...
 
-		database_wiiu.UpdateUserLastOnlineTime(pid, lastOnline)
-		sendUserWentOfflineWiiUNotifications(packet.Sender())
+		if platform == 1 {
+			database_wiiu.UpdateUserLastOnlineTime(pid, lastOnline)
+			friends_wiiu.SendUserWentOfflineNotifications(packet.Sender())
+		} else if platform == 2 {
+			database_3ds.UpdateUserLastOnlineTime(pid, lastOnline)
+			friends_3ds.SendUserWentOfflineNotificationsGlobally(packet.Sender())
+		}
```
Comment (jonbarrow) [inline nex_server.go]: These platforms should probably be an enum or global variables for easier reading. Such as `WUP`/`CTR` (since variables can't start with a number for the 3DS)

File: register_ex.go [hunk 9]
```diff
@@ -2,15 +2,33 @@ package main
...
+
+	if loginData.TypeName() == "NintendoLoginData" {
+		user.Platform = 1 // Platform is Wii U
+
+		database_wiiu.UpdateUserLastOnlineTime(pid, lastOnline)
+	} else if loginData.TypeName() == "AccountExtraInfo" {
+		user.Platform = 2 // Platform is 3DS
+
+		database_3ds.UpdateUserLastOnlineTime(pid, lastOnline)
+	}
```
Comment (jonbarrow) [inline register_ex.go]: Platform check should use an enum or global variables?

**Review by jonbarrow — CHANGES_REQUESTED** [review]

File: assign_nex_protocols.go (hunk 1 above)
Comment (hauntii) [reply assign_nex_protocols.go]: Not actually sure, don't remember the reason. Its been renamed back.

File: 3ds/update_presence.go (hunk 7 above)
Comment (hauntii) [reply 3ds/update_presence.go:13]: No clue, only FF FF FF FF, which is all flags. No other ones seem to be useful, as this is only for removing all of the data to hide what the player is doing.

File: 3ds/notifications.go (hunk 3 above)
Comment (hauntii) [reply 3ds/notifications.go]: Split them into different files.

File: 3ds/update_presence.go (hunk 7 above)
Comment (jonbarrow) [reply 3ds/update_presence.go:13]: Okay. I was asking because if other flags were known then it would likely be better to also use an enum or global variable for those as well. But if not then hard coding it should be fine

**Comment by jonbarrow** [comment]
Looks awesome!
